File: py/yeriasdk/views/action_grid_view.py
"""
ActionGridView - A view for displaying actions in a grid format
"""

from typing import Optional
from datetime import datetime

from .base_action_view import BaseActionView
from ..types.models import ActionConfig
from ..errors.exceptions import InvalidParameterError


class ActionGridView(BaseActionView):
    """View for displaying actions in a grid"""

    def __init__(self, view_id: str, title: str, process_id: Optional[str] = None):
        super().__init__(
            {
                "id": view_id,
                "type": "ActionGrid",
                "process_id": process_id,
                "metadata": {
                    "version": "1.0.0",
                    "created_at": datetime.now(),
                },
            }
        )

        self.content = {
            "title": title,
            "intro": "",
            "actions": self._actions,  # Reference to actions from BaseActionView
            "columns": 2,
            "spacing": 16,
        }

    def set_title(self, title: str) -> "ActionGridView":
        """Set the action grid title"""
        self.content["title"] = title
        return self

    def set_intro(self, intro: str) -> "ActionGridView":
        """Set the introduction text displayed before the action grid"""
        return self._set_intro_text("intro", intro)

    def set_columns(self, columns: int) -> "ActionGridView":
        """Set the number of columns"""
        if columns < 1 or columns > 6:
            raise InvalidParameterError(
                "columns", columns, "Columns must be between 1 and 6"
            )
        self.content["columns"] = columns
        return self

    def set_spacing(self, spacing: int) -> "ActionGridView":
        """Set spacing between elements"""
        if spacing < 0:
            raise InvalidParameterError("spacing", spacing, "Spacing must be non-negative")
        self.content["spacing"] = spacing
        return self

    def _sync_actions_to_content(self) -> None:
        """Sync actions array to content (required by BaseActionView)"""
        self.content["actions"] = [
            {
                "code": action.code,
                "title": action.title,
                "desc": action.desc,
                "thumbnail": action.thumbnail,
                "disabled": action.disabled if action.disabled else False,
                "metadata": action.metadata,
            }
            for action in self._actions
        ]

    def add_action(
        self,
        code: str,
        title: str,
        description: Optional[str] = None,
        thumbnail: Optional[str] = None,
        disabled: bool = False,
        metadata: Optional[dict] = None,
    ) -> "ActionGridView":
        """Add an action and sync to content"""
        super().add_action(code, title, description, thumbnail, disabled, metadata)
        self._sync_actions_to_content()
        return self

    def remove_action(self, action_code: str) -> bool:
        """Remove an action and sync to content"""
        result = super().remove_action(action_code)
        if result:
            self._sync_actions_to_content()
        return result

    def update_action(self, action_code: str, updates: dict) -> bool:
        """Update an action and sync to content"""
        result = super().update_action(action_code, updates)
        if result:
            self._sync_actions_to_content()
        return result

    def get_columns(self) -> int:
        """Get the number of columns"""
        return self.content.get("columns", 2)

    def get_spacing(self) -> int:
        """Get the spacing"""
        return self.content.get("spacing", 16)

    def has_actions(self) -> bool:
        """Check if the grid has actions"""
        return len(self._actions) > 0

File: js/src/core/action-grid-view.ts
import { BaseActionView } from './base/base-action-view';
import { ActionConfig } from '../types';
import { InvalidParameterError } from '../errors';

export interface ActionGridContent {
    title?: string;
    intro?: string;
    actions: ActionConfig[];
    columns?: number;
    spacing?: number;
}

/**
 * Builds an SGUI `ActionGrid` view — actionable items laid out in a grid, each
 * with a code, title, optional description/thumbnail, and navigation.
 *
 * Builder methods: `setTitle`, `setIntro`, `setColumns` (1-6), `setSpacing`,
 * and the action family inherited from BaseActionView (`addAction`,
 * `removeAction`, `updateAction`, `hasActions`), overridden here to keep the
 * serialized `content.actions` in sync.
 *
 * Extends {@link BaseActionView}. Created via the YeriaApp/YeriaUI factory,
 * populated with these builders, then serialized and signed into a v3 envelope
 * by serve().
 */
export class ActionGridView extends BaseActionView {

    static fromJson(json: Record<string, unknown>): ActionGridView {
        return ActionGridView.fromJsonAs(ActionGridView, 'ActionGrid', json);
    }
    constructor(viewId: string, title: string, processId?: string) {
        super({
            id: viewId,
            type: 'ActionGrid',
            processId,
            metadata: {
                version: '1.0.0',
                createdAt: new Date()
            }
        });

        this.content = {
            title: title,
            intro: '',
            actions: this.actions,  // Reference to BaseActionView's actions
            columns: 2,
            spacing: 16
        } as ActionGridContent;
    }

    /**
     * Sets the action grid title
     */
    setTitle(title: string): this {
        (this.content as ActionGridContent).title = title;
        return this;
    }

    /**
     * Sets the action grid intro
     */
    setIntro(intro: string): this {
        return this.setIntroText('intro', intro);
    }

    /**
     * Sets the number of columns
     */
    setColumns(columns: number): this {
        if (columns < 1 || columns > 6) {
            throw new InvalidParameterError('columns', columns, 'Columns must be between 1 and 6');
        }
        (this.content as ActionGridContent).columns = columns;
        return this;
    }

    /**
     * Sets the spacing between items
     */
    setSpacing(spacing: number): this {
        if (spacing < 0) {
            throw new InvalidParameterError('spacing', spacing, 'Spacing must be non-negative');
        }
        (this.content as ActionGridContent).spacing = spacing;
        return this;
    }

    /**
     * Synchronize actions array to content (required by BaseActionView)
     */
    protected syncActionsToContent(): void {
        (this.content as ActionGridContent).actions = this.actions;
    }

    /**
     * Overrides addAction to keep content in sync
     */
    override addAction(
        code: string,
        title: string,
        description?: string,
        thumbnail?: string,
        disabled: boolean = false,
        metadata?: Record<string, unknown>
    ): this {
        super.addAction(code, title, description, thumbnail, disabled, metadata);
        this.syncActionsToContent();
        return this;
    }

    /**
     * Overrides removeAction to keep content in sync
     */
    override removeAction(actionCode: string): boolean {
        const result = super.removeAction(actionCode);
        if (result) {
            this.syncActionsToContent();
        }
        return result;
    }

    /**
     * Overrides updateAction to keep content in sync
     */
    override updateAction(actionCode: string, updates: Partial<ActionConfig>): boolean {
        const result = super.updateAction(actionCode, updates);
        if (result) {
            this.syncActionsToContent();
        }
        return result;
    }

    /**
     * Gets the number of columns
     */
    getColumns(): number {
        return (this.content as ActionGridContent).columns || 2;
    }

    /**
     * Gets the spacing
     */
    getSpacing(): number {
        return (this.content as ActionGridContent).spacing || 16;
    }

    /**
     * Checks whether the grid has any actions
     */
    hasActions(): boolean {
        return this.actions.length > 0;
    }
}
